business/suggestionService.js

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation and others. Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+const { get, set, sortBy, filter, concat, last, isEmpty } = require('lodash')
+
+class SuggestionService {
+  constructor(definitionService, definitionStore) {
+    this.definitionService = definitionService
+    this.definitionStore = definitionStore
+  }
+
+  /**
+   * Get suggestions for the given coordinates.
+   *
+   * @param {EntityCoordinates} coordinates - The entity for which we are looking for suggestions
+   * @returns {Suggestion} A set of suggested fixes for the definition at the given coordinates.
+   * `null` is returned if no such coordinates are found.
+   */
+  async get(coordinates) {
+    const related = await this._getRelatedDefinitions(coordinates)
+    if (!related) return null
+    const baseSuggestion = this._createBaseSuggestion(coordinates)
+    // create a struct with the strings in it and iterate instead
+    const promises = [
+      await this._collectSuggestionsForField(related, baseSuggestion, 'licensed.declared'),
+      await this._collectSuggestionsForField(related, baseSuggestion, 'described.sourceLocation'),
+      await this._collectSuggestionsForField(related, baseSuggestion, 'described.releaseDate')
+    ]
+    const result = await Promise.all(promises)
+    return last(result.filter(res => !isEmpty(res)))
+  }
+
+  /**
+   * Get related definitions for the given coordinates.
+   * Related here means other definitions for the same component with different revisions
+   *
+   * @param {EntityCoordinates} coordinates - The entity we are looking for related defintions to
+   */
+  async _getRelatedDefinitions(coordinates) {
+    const related = await this.definitionStore.list(coordinates.asRevisionless(), 'definitions')
+    // If the related array only has one entry then return early
```

Comment:
why return early? the comment does not add anything to over the code.

Reply:
The related array at this stage contains all the found definitions in the store, including the current one that we are passing.
So if the related array contains only one element, it is the same one that we have passed, and so we can exit early here.

But we have some questions on this approach, because your first cut uses two arrays (before and after) to know what are the 3 revisions after the passed one, and the 3 past before.
We could maybe improve it using some `lodash` methods like `take`, but we are asking if this is really what we expect in terms of returned data or if you have any other thoughts?

Reply:
I think that the shape of the return suggestion data will change considerably over the next while. In general the data coming back from the service should be well structured as suggestions. In the fullness of time we'll be getting suggestions from all over so will get more "sophisticated" than a list of all versions of the component.

I'm fine with taking some simple approach now. If you want to do that, please note the reason for returning early in the comment.

Does that cover your questoins?

Reply:
I will extend the comment in the code.
But from your answer, which is pretty reasonable, i'm still not able to understand why to get the information only from the 3 more recent and the 3 past before the current version, instead of taking, as an example, the latest 6 versions, without considering what's before or after the current revision.

Reply:
Oh, I just picked numbers. 3, 5, 6, ... as for the before and after, seemed reasonable that the user would consider examples that "surround" the definition in question. Not wed to this idea. Similarly, we should not get wed to any other approach at this point. They are very likely to be wrong. I just picked one that "seemed interesting". We need to pick something and try it.

Reply:
I see, we can go with this and see what happens.